src/Algorithms/CARTcommon.cpp

```diff
@@ -35,63 +35,40 @@
 #include <shark/Core/Math.h>
 #include <shark/Algorithms/Trainers/RFTrainer.h>
 
+#include <limits>
+
 namespace shark {
 namespace detail{
 namespace cart {
 
-RealVector hist(ClassVector const& countVector)
-{
-
-	RealVector histogram(countVector.size(),0.0);
-
-	std::size_t totalElements = 0;
-
-	for (std::size_t i = 0, s = countVector.size(); i<s; ++i){
-		histogram(i) = countVector[i];
-		totalElements += countVector[i];
-	}
-	histogram /= totalElements;
-
-	return histogram;
-}
 
 ///Calculates the Gini impurity of a node. The impurity is defined as
 ///1-sum_j p(j|t)^2
 ///i.e the 1 minus the sum of the squared probability of observing class j in node t
 double gini(ClassVector const& countVector, std::size_t n)
 {
-	if(!n) return 1;
-
-	double res = 0.;
-	n *= n;
-	for(auto const& i: countVector){
-		res += sqr(i)/ static_cast<double>(n);
-	}
-	return 1-res;
+	if(!n) return 1.;
+	return 1.-sum(sqr(countVector/double(n)));
 }
 
 // ME = 1-max(count)/n
 double misclassificationError(ClassVector const& countVector, std::size_t n)
 {
 	if(!n) return 1.;
-	auto m = countVector[0];
-	for(auto c: countVector){
-		m = std::max(m,c);
-	}
-	return 1-m/n;
+	double m = max(countVector);
+	return 1.-m/n;
 }
 
-// CE = - sum_j(count[j] log(count[j]/n))/n
+// CE = - sum_j(count[j]/n * log(count[j]/n))
 double crossEntropy(ClassVector const& countVector, std::size_t n)
 {
-	return - sum<double>(0,n,[&](std::size_t i){
-		if(!countVector[i]) return 0.;
-		return countVector[i]*std::log(countVector[i]/n);
-	})/n;
+	if(!n) return std::numeric_limits<double>::infinity();
```

Comment:
Odd, I was sure I tested this and got a similar (ie. only slightly different) result back as previously with `crossEntropy` in the end I figured `blas::log` was doing something funky to avoid all the infinities, and the fact I was getting numbers back at all threw me off the scent, but you're absolutely right, we need to filter out the zeroes.

As for the return of infinity.. I'm testing `n` not `p`, so technically as `n` → +0,  `p` → infinity and `-p*log(p)` → &minus;infinity, meaning I ought to return the negative infinity. Admittedly though, I was coming at it from a somewhat more brutish point of view: We're measuring impurity, high impurity is bad, if `n==0` that is also bad, thus return a disincentive.. and speaking of... if `n` → &minus;0 then `-p*log(p)` → +infinity, so maybe it's kind of alright?